### PR TITLE
(WORK IN PROGRESS) Add a post about Carina, SSL, and Docker

### PR DESCRIPTION
**Pull request by larsbutler**
Hi! This post is pretty much, except for some mechanical issues. I'm trying to figure how to get my table of contents links to the section fragments to work properly, but I'm not having much luck. I could use some syntactical assistance there. =)

Other feedback is welcome, too. This is my first contribution.

Also, is Christmas Eve a bad date to publish? What I mean is, is this an automated procedure or a manual one?

### CODE REVIEW

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+- [Conclusion](#conclusion)
+
+## Introduction<a href="introduction"></a>
+
+It's almost 2016, so I shouldn't need to convince you that making your web
+application as secure as possible is an absolute necessity. Security is a
+multi-faceted topic, but probably the biggest application security concerns
+revolve around deployment and configuration of infrastructure. Internally,
+that means that secrets, such as passwords and private keys, are stored and
+handled in a secure way and that network access to resources, such as virtual
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... secure way, and that network access ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+channel, and that usually means that you'll need to set up SSL (Secure Sockets
+Layer) on your web service in order to communicate over secure HTTP (HTTPS). In
+terms of getting your application securely into production, there's a lot to
+think about. But if you're a software developer like me, you probably want
+spend to _more_ time thinking about building the next feature for your app and
+_less_ time thinking about how to deploy it.
+
+In this post, I'll show you what I found to be the path of least resistance in
+getting a secure web application up and running on the Rackspace Cloud. I'll
+first take you quickly through the process of creating an SSL certificate, then
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... SSL certificate, and then ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+- The certificate itself, typically saved as a `.crt` file. This will be given
+  to you after a successful certificate request.
+- The intermediate certificate chain, or Certificate Authority (CA) bundle.
+  This will be given to you along with your certificate.
+
+You will need all three files to set up SSL on your web application.
+
+## Environment Setup<a href="environment-setup"></a>
+
+Before we can start hacking on things, we'll need to install some tools. Login
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... some tools. Log in ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+to your Linux or OSX machine and install the Docker Version Manager (dvm):
+
+    wget https://download.getcarina.com/dvm/latest/install.sh
+    sh install.sh
+
+The output of this command will tell you to `source /home/me/.dvm/dvm.sh`
+(where `/home/me` is your home directory) and then add that command to your
+`.bashrc` or `.bash_profile. Let's do that:
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... or `.bash_profile`. Let's ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+    echo "/home/me/.dvm/dvm.sh" >> ~/.bashrc
+
+## Let's Do This<a href="lets-do-this"></a>
+
+Okay, now let's get to the fun part. Let's create a simple web application and
+run it on Carina. First, [login to Carina](https://app.getcarina.com/app/login).
+This will drop you immediately into
+the Carina control panel. Click on `Add Cluster...` and give it a name. You can
+call it whatever you want. For example, you could name your cluster after the
+FQDN for your site, such as `secure-mydomain-com` (the name can only contain
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... (Note: the name can ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+  8080, so we need to expose that port to the outside world.
+- `ENTRYPOINT ["./main"]`: The command to execute to start the web service.
+
+Now it's time to build our Docker container image and deploy it:
+
+    source docker.env
+    dmv use
+
+`source docker.env` sets up environment variables we'll need in order to talk
+to the Docker service running on Carina, and `dvm use` will set up a compatible
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... and `dvm use` sets up a ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+
+Now let's get the ServiceNet IP for the node:
+
+    $ docker run --net=host \
+        --env constraint:node==885bbfe5-5fd1-4797-97d1-2ae199906afb-n1 \
+        racknet/ip \
+        service ipv4
+    10.176.229.17
+
+Copy the resulting IP address, then spin up two containers. The command you
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... address, and then spin ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+need is similar to the `docker run` commands we used before, but with a couple
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ...  before but with a couple ...  [ ed note: or ... before, but there are  a couple ... ]

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+    docker run --detach --publish 10.176.229.17::8080 \
+        --env constraint:node=885bbfe5-5fd1-4797-97d1-2ae199906afb-n1 \
+        secure-mydomain-com
+
+Run this command a second time so we have two containers.
+
+So what's going on in this command? There are two main things to focus on here:
+
+- `--publish 10.176.229.17::8080`: The format of the `--publish` argument is
+  `ip:hostPort:containerPort`. So what this means is, publish the service
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ...  ip:hostPort:containerPort`. This means: publish ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+  running on `8080` in the container to the ServiceNet IP on the host, and use
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: .. the host and use ..

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+    $ docker ps
+    CONTAINER ID        IMAGE                 COMMAND             CREATED             STATUS              PORTS                           NAMES
+    7ac35cf8c9d2        secure-mydomain-com   "./main"            3 minutes ago       Up 28 seconds       10.176.229.17:32771->8080/tcp   885bbfe5-5fd1-4797-97d1-2ae199906afb-n1/prickly_panini
+    8398b4e45b1c        secure-mydomain-com   "./main"            3 minutes ago       Up 31 seconds       10.176.229.17:32768->8080/tcp   885bbfe5-5fd1-4797-97d1-2ae199906afb-n1/condescending_tesla
+
+Take note of the host IP address and port for each container. We need to add
+these to our load balancer. Head over to you Rackspace Cloud control panel and
+open up your load balancer configuration. In the `Nodes` configuration, remove
+the two old nodes (by clicking the gear and selecting
+`Remove from load balancer...`), then add two entries for our new containers,
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... balancer...`), and then add two ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+     bundle file isn't adequate, you can specify an alternate file
+     using the --cacert option.
+    If this HTTPS server uses a certificate signed by a CA represented in
+     the bundle, the certificate verification probably failed due to a
+     problem with the certificate (it might be expired, or the name might
+     not match the domain name in the URL).
+    If you'd like to turn off curl's verification of the certificate, use
+     the -k (or --insecure) option.
+
+Basically what this error is telling us, we should only be accessing our web
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: Basically, this error tells us that we should only ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+
+Try it out in a browser, too, and you should get a "green light" on your secure
+connection!
+
+![Browser - Secure HTTP]({% asset_path 2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker/8-https-secure.png %})
+
+## Conclusion<a href="conclusion"></a>
+
+In this tutorial, I've tried to give a simple and realistic example of
+deploying a web application with SSL, but without leaving out important
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... with SSL without leaving ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+details. Naturally, a real web application is going to be more complex than a
+few containers running a "hello, world" program behind a load balancer; at some
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... a load balancer. At some ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 15]
```diff
@@ -0,0 +1,614 @@
...
+point, so you will likely want to add a datastore, a cache, and/or other
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: ... point, you may want to ...

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+services. The [Carina documentation](https://getcarina.com/docs/#tutorials)
+already has a bunch of great tutorials for containerizing services like MySQL,
+Apache, and MongoDB, so check them out! Happy hacking.
```
Comment (catlook) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: [ ed note: great tutorial!! Use any edits of mine you like and discard the others :) 
Also, I think publishing on xmas eve should be fine. - Cat ]

**Comment by everett-toews** [comment]
Hi Lars. After you've made your edits, can you please move this post to https://github.com/getcarina/getcarina.com/tree/master/_posts as it's Carina focused?

You'll also find assets/img in a similar place over there too.

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown (hunk 16 above)
Comment (larsbutler) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: Thanks for the thorough review, Cat! This gives it some nice polish.

**Comment by larsbutler** [comment]
@everett-toews Sure thing. I had considered submitting it to the Carina blog initially; I'll move it over there.

File: _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown (hunk 4 above)
Comment (larsbutler) [inline _posts/2015-12-24-deploying-secure-web-applications-rackspace-cloud-carina-docker.markdown]: Ah, good catch. Thanks!

**Comment by larsbutler** [comment]
Edits made. Closing this and re-submitting to the Carina blog.

**Pull request closed without merging**